packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,131 @@
+import { GetSigners } from "src/Safe/GetSigners.sol";
+import { Enum } from "safe-contracts/common/Enum.sol";
+import { ISemver } from "src/universal/ISemver.sol";
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+contract LivenessGuard is ISemver, BaseGuard {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /// @notice Emitted when a new set of signers is recorded.
+    /// @param signers An arrary of signer addresses.
+    event SignersRecorded(bytes32 indexed txHash, address[] signers);
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice The safe account for which this contract will be the guard.
+    Safe public immutable safe;
+
+    /// @notice A mapping of the timestamp at which an owner last participated in signing a
+    ///         an executed transaction.
+    mapping(address => uint256) public lastLive;
+
+    /// @notice An enumerable set of addresses used to store the list of owners before execution,
+    ///         and then to update the lastSigned mapping according to changes in the set observed
+    ///         after execution.
+    EnumerableSet.AddressSet private ownersBefore;
+
+    /// @notice Constructor.
+    /// @param _safe The safe account for which this contract will be the guard.
+    constructor(Safe _safe) {
+        safe = _safe;
+    }
+
+    /// @notice We use this post execution hook to compare the set of owners before and after.
+    ///         If the set of owners has changed then we:
+    ///         1. Add new owners to the lastSigned mapping
+    ///         2. Delete removed owners from the lastSigned mapping
+    function checkAfterExecution(bytes32, bool) external {
+        // Get the current set of owners
+        address[] memory ownersAfter = safe.getOwners();
+
+        // Iterate over the current owners, and remove one at a time from the ownersBefore set.
+        for (uint256 i = 0; i < ownersAfter.length; i++) {
+            // If the value was present, remove() returns true.
+            if (ownersBefore.remove(ownersAfter[i]) == false) {
+                // This address was not already an owner, add it to the lastSigned mapping
+                lastLive[ownersAfter[i]] = block.timestamp;
+            }
+        }
+        // Now iterate over the remaining ownersBefore entries. Any remaining addresses are no longer an owner, so we
+        // delete them from the lastSigned mapping.
+        for (uint256 j = 0; j < ownersBefore.length(); j++) {
+            address owner = ownersBefore.at(j);
+            delete lastLive[owner];
+        }
+        return;
+    }
+
+    /// @notice Records the most recent time which any owner has signed a transaction.
+    /// @dev This method is called by the Safe contract, it is critical that it does not revert, otherwise
+    ///      the Safe contract will be unable to execute transactions.
```

Comment:
I need to check to make sure this is called once before transaction execution and not on each time that a signature is sent

Reply:
Can you add a comment that this is called in `execTransaction` before making the call? It could be good to note either here or at the top level natspec that `execTransaction` can only be called successfully if there are enough signatures. The reason why I suggest at the contract level natspec is because both `checkTransaction` and `checkAfterExecution` operate from within `execTransaction`

Reply:
4987e4a59ddb8b94a061dcb9fcb05c7e0dba27d8